pkg/zstdchunked/storage.go

```diff
@@ -0,0 +1,983 @@
+	maps := make(map[string]map[string]*FileMetadata)
+
+	for layerID, v := range layersMetadata {
+		r := make(map[string]*FileMetadata)
+		for i := range v {
+			r[v[i].Digest] = &v[i]
+		}
+		maps[layerID] = r
+	}
+	return maps
+}
+
+func getLayersCache(store storage.Store) (map[string][]FileMetadata, map[string]string, error) {
+	allLayers, err := store.Layers()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	layersMetadata := make(map[string][]FileMetadata)
+	layersTarget := make(map[string]string)
+	for _, r := range allLayers {
+		manifestReader, err := store.LayerBigData(r.ID, bigDataKey)
+		if err != nil {
+			continue
+		}
+		defer manifestReader.Close()
+		manifest, err := ioutil.ReadAll(manifestReader)
+		if err != nil {
+			return nil, nil, err
+		}
+		var toc TOC
+		if err := json.Unmarshal(manifest, &toc); err != nil {
+			continue
+		}
+		layersMetadata[r.ID] = toc.Entries
+		target, err := store.DifferTarget(r.ID)
+		if err != nil {
+			return nil, nil, err
+		}
+		layersTarget[r.ID] = target
+	}
+
+	return layersMetadata, layersTarget, nil
+}
+
+func MakeZstdChunkedDiffer(ctx context.Context, store storage.Store, blobSize int64, annotations map[string]string, iss ImageSourceSeekable) (*ChunkedZstdDiffer, error) {
+	manifest, err := ReadZstdChunkedManifest(iss, blobSize, annotations)
+	if err != nil {
+		return nil, err
+	}
+	layersMetadata, layersTarget, err := getLayersCache(store)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ChunkedZstdDiffer{
+		stream:         iss,
+		manifest:       manifest,
+		ctx:            ctx,
+		layersMetadata: layersMetadata,
+		layersTarget:   layersTarget,
+	}, nil
+}
+
+func findFileInOtherLayers(file FileMetadata, root string, dirfd int, layersMetadata map[string]map[string]*FileMetadata, layersTarget map[string]string, missingDirsMode os.FileMode) (*os.File, int64, error) {
+	// this is ugly, needs to be indexed
+	for layerID, checksums := range layersMetadata {
+		m, found := checksums[file.Digest]
+		if !found {
+			continue
+		}
+
+		source, ok := layersTarget[layerID]
+		if !ok {
+			continue
+		}
+
+		srcDirfd, err := unix.Open(root, unix.O_RDONLY, 0)
+		if err != nil {
+			continue
+		}
+		defer unix.Close(srcDirfd)
+
+		srcFile, err := openFileUnderRoot(m.Name, source, srcDirfd, unix.O_RDONLY, 0)
+		if err != nil {
+			continue
+		}
+		defer srcFile.Close()
+
+		srcPath := fmt.Sprintf("/proc/self/fd/%d", srcFile.Fd())
+
+		dstFile, written, err := copyFileContent(srcPath, file.Name, root, dirfd, missingDirsMode, 0)
+		if err != nil {
+			continue
+		}
+		return dstFile, written, nil
+	}
+	return nil, 0, nil
+}
+
+// findFileOnTheHost checks whether the requested file already exist on the host and copies the file content from there if possible.
+// It is currently implemented to look only at the file with the same path.  Ideally it can detect the same content also at different
+// paths.
```

Comment:
Not really related to this PR, but in OpenVZ we used to have a filesystem which
 - maintained file digest in xattrs (recalculated and updated by the kernel on close() after write);
 - in read-only open() or mmap(), looked for the file with the same checksum in "master" directory and served it from the master.
 
It's a bit more complicated than that (a daemon which create "master" files, copy-on-write etc) but this helped a lot with container density.

Reply:
I thought about improving `mlocate` to also store the file content digest on `updatedb`, but that would be a breaking change in its DB.

Reply:
This is of course very similar to how ostree works, although the digest here covers the more of the inode and not just the contents (xref https://github.com/ostreedev/ostree/issues/2259).

Reply:
and ostree integration is definitely something I'd like to have in future!  And that was something I was planning already long time ago but never found the time to implement when I've added https://github.com/ostreedev/ostree/pull/1443

Moving containers/storage to store the checksum for each file will make it finally possible (and we'll need a better store than a list of JSON files to store the files and iterate each of them)